### DISCUSSION TITLE
Overview of history of starship init

### DISCUSSION THREAD

**Discussion by chipbuster** [starship/starship#3491] [General]
Starship's initialization procedure is built out of quite a few pieces: the idea of shipping the initialization in the binary itself, the two-phase initialization which makes it so that the entire initscript can be evaluated in a single line, and multiple other subsystems.

Many of these mechanisms were tuned via trial-and-error, and we historically haven't kept a log of all the issues we've hit. I went back through the repo history and looked at all PRs that I thought satisfied the following criteria:

- Were made to the init system or init scripts
- Fixed a bug (i.e. were not feature changes or refactorings) that could potentially cause the init system to completely break. Issues with quoting and variable capture (e.g. #3088, #2100) were mostly ignored.

These are the PRs I ended up looking at:

| PR  | Linked Against | Summary |
| ------------- | ------------- | --- | 
| #2848  | #2767 | Issues with how the name of the binary is quoted in init | 
| #1541  | None  | `PROMPT_COMMAND` can cause infinite recursion with `starship_precmd` |
| #1185 | #954 | Preserve shell state by using bash stack to store function lists |
| #890 | #842 | Discovered issue with interaction with Bash-it (eventually fixed on their end) |
| #799 | None | PROMPT_COMMAND needs to be terminated, else other frameworks can append badly |
| #696 | None | `starship_precmd` cannot go at the front of `PROMPT_COMMAND` |
| #336 | #330 | `PROMPT_COMMAND` cannot be overwritten (else other frameworks break) |
| #278 | None | `/dev/stdin` does not exist on all systems--need to use caution |
| #241 | None | termux does not have `/dev/stdin` |
| #224 | #220 | Use runtime information to modify initscripts to allow relative-path calls to init |
| #213 | #196 | Double-bracket conditionals do not exist in older versions of ZSH |
| #180 | None | Cannot rely on some variables being defined |
| #106 | #100 | Allow invocations of starship with absolute paths |

In addition, we have support issues with older versions of `bash`: because MacOS
is frozen on bash 3.2 (released 2006-10-11), there are many modern features that
we cannot always rely on. Since MacOS has moved to `zsh` being the default shell,
we can likely relax this condition in the future, but at the moment, we try to
support all versions of bash from 3.2 to current.

Here are the conclusions from each individual PR (in reverse chronological order).
I will summarize everything at the end.

## PR #106, #224

Starship used to call itself just by `starship` in the init files. 
This is incorrect because the user may specify non-PATH paths to 
starship, or just have multiple copies of starship in the PATH but
not want to use the first.

These PRs added support for absolute and relative path invocations
of starship. As long as `starship` can be found when `starship init` 
is called and the binary does not move, starship should continue to work.

## PR #180

A dumb mistake on my part. Even when some variables are used to control shell
behavior, we cannot always rely on them being defined.

## PR #213

This fixed an issue where double-bracket conditionals (e.g. `if [[ $VAR ]]`)
completely broke ZSH. This *seems* to be because early versions of ZSH did not
support this syntax.

Features in shells cannot always be relied upon to be present if the shell has
a long lineage. This is likely most relevant for long-lived popular shells
like bash, zsh, and tcsh.

## PR #241, #278

The bash-3.2 fallback init code uses /dev/stdin to get around rules about
sourcing with process substitution. Unfortunately, many of our platforms (incl.
Termux and Windows) do not have `/dev/stdin`, meaning we have to carefully
control when this fallback is called.

## PR #336

This fixed an issue caused by setting `PROMPT_COMMAND` directly--we cannot do
this (same with `DEBUG` traps) since other frameworks hook this to do their own
setup.

## PR #696, #799

Starship shares the `PROMPT_COMMAND` hook with other programs, which may attempt
to do things such as append to it without including semicolons. There may also
just be [no other programs](https://github.com/starship/starship/pull/696#discussion_r356920660).

In all cases, starship needs to function correctly or be a good citizen and
attempt to avoid breaking other scripts as much as possible.

## PR #1185

A very clever change by MunifTanjim which sidesteps many of the issues we've seen
above (though the lessons learned from those issues should be kept in mind).
Replaces the current `PROMPT_COMMAND`, but preserves the old one in a variable.
Then, once starship has finished running its `precmd`, it executes the variable
storing the old `PROMPT_COMMAND`.

## PR #1541

Fixes an infinite recursion that can occur if `starship_precmd` gets into 
`PROMPT_COMMAND` more than once.

## PR #2848

Changes shell quoting of the path to the starship binary.

# Takeaways

- Modern shell support seems to mostly work as expected. Support for shells with
  long histories (primarily bash, zsh, and possibly tcsh), is made trickier by
  the fact that older versions do not support features that new ones do, or have
  backwards-incompatible changes. NOTE: Fish will also exhibit this behavior
  as time goes on.

- We cannot rely on platform-specific files existing in the general case.

- Bash init is a pain.

# Use cases we support

(i.e. Use cases we have made changes to to get working in the past, and should not break without making an explicit decision to do so)

- Running bash-3.2 on a platform which has `/dev/stdin`
- Running bash-4.1+ on *any* platform, including termux and Windows
- Running zsh-4.0+ on any platform 
- Running either fish 2.x/3.x
- All the other shells we have official support for

**Comment by matchai** [comment]
Excellent roundup, @chipbuster!
I was starting to feel like our decisions around the init script were being lost to time. It's excellent having all this valuable history in one place! 🙌
